scripts/cmd/build.js

```diff
@@ -138,3 +141,44 @@ async function clean(outdir) {
 		onlyFiles: true,
 	});
 }
+
+/** 
+  * Contextual `define` values to statically replace in the built JS output.
+  * Available to all packages, but mostly useful for CLIs like `create-astro`.
+  */
+async function getDefinedEntries() {
+	const define = {
+		/** The current version (at the time of building) for the current package, such as `astro` or `@astrojs/sitemap` */
+		PACKAGE_VERSION: await getInternalPackageVersion('./package.json'),
+		/** The current version (at the time of building) for `astro` */
+		ASTRO_VERSION: await getInternalPackageVersion(new URL('../../packages/astro/package.json', import.meta.url)),
+		/** The current version (at the time of building) for `@astrojs/check` */
+		ASTRO_CHECK_VERSION: await getExternalPackageVersion('@astrojs/check'),
+		/** The current version (at the time of building) for `typescript` */
+		TYPESCRIPT_VERSION: await getExternalPackageVersion('typescript'),
+	}
+	return Object.entries(define);
+}
+
+async function readPackageJSON(path) {
+	return await fs.readFile(path, { encoding: 'utf8' }).then((res) => JSON.parse(res));
+}
+
+async function getInternalPackageVersion(path) {
+	return readPackageJSON(path).then(res => res.version);
+}
+
+const DEPENDENCY_KEY_RE = /dependencies$/i;
+async function getExternalPackageVersion(packageName) {
+	const { stdout } = await execa('pnpm', ['list', packageName, '-r', '--json', '--depth', '-1'], { encoding: 'utf8', stdio: 'overlapped' });
+	if (!stdout) return;
+	const matches = JSON.parse(stdout);
+	for (const pkg of matches) {
+		for (const [key, value] of Object.entries(pkg)) {
+			if (DEPENDENCY_KEY_RE.test(key) && value[packageName] !== undefined) {
+				const { version } = value[packageName];
+				return version;
+			}
+		}
+	}
+}
```

Comment:
For `@astrojs/check` and `typescript`, maybe we can read them from the root `package.json` instead. I think it'll be safer and more stable than running through `execa`. Also, I think we should error if we're unable to find any version, otherwise the defined vars won't get replaced and they fallback to `undefined`.

Reply:
Yep, great suggestion. Refactored to read from the workspace root instead!

As for the missing behavior, I also updated that to throw a more helpful error.